ENH: Transform composition

Aakanksha-Rana opened the pull request:
## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to change)

## Summary
<!--- What does your code do? -->

## Checklist
<!--- Please, let us know if you need help-->
- [x] All tests passing
- [x] I have added tests to cover my changes
- [ ] I have updated documentation (if necessary)
- [x] My code follows the code style of this project
(we are using `black`: you can `pip install pre-commit`,
run `pre-commit install` in the `pydra` directory
and `black` will be run automatically with each commit)

## Acknowledgment
- [x] I acknowledge that this contribution will be available under the Apache 2 license.

satra commented:
let's use this PR to discuss and implement the design of a data augmentation composition pipeline similar to Monai.

Aakanksha-Rana commented:
transform = ComposeTransform([
    Combine(),              #  merges input and ground truth images along channel-dimension
    Transform1(),
    Transform2(),
     ...
    Split(,),  # split back into input and ground truth
    Transform3().     # e.g. Normalization[input, output]) should be done separately

This could be our design. Thoughts?  
The compose function is similar to Monai (if I say originally pytorch...). But combine and split doesn't exist in monai. 

We may also discuss what all transformations we need in this framework, where input and target are to be handled together/separately.

satra commented:
thanks @Aakanksha-Rana - regarding split and combine, is the idea that we create a single tensor to operate on? 

we may want to augment as part of a tensorflow graph as opposed to outside. 

- how about thinking about this through the lens of tensorflow ops? would this work in that setting?
- what about looking at the scikit-learn transform API, would that be useful?
- what about multimodal inputs and outputs?

kaczmarj commented:
>we may want to augment as part of a tensorflow graph as opposed to outside

I agree, we should do this as part of the tensorflow graph, preferably inside a `tf.data.Dataset` operation ([`dataset.map`](https://www.tensorflow.org/api_docs/python/tf/data/Dataset#map)). Those operations can be parallelized, which can be very useful for augmentations that take significant time. I have implemented a few operations with tensorflow to be used in a tf.data dataset.

>what about multimodal inputs and outputs

multimodal inputs are a good argument for not concatenating features and labels in the channel dimension. I would prefer if we could keep features and labels separate but create transformation compositions that operate on both of them. i wonder what the scikit-learn api has done about this

Aakanksha-Rana commented:
> thanks @Aakanksha-Rana - regarding split and combine, is the idea that we create a single tensor to operate on?
A combine step only when we want same operations, like rotate, crop.; yes one tensor for that. 

> we may want to augment as part of a tensorflow graph as opposed to outside.

>     * how about thinking about this through the lens of tensorflow ops? would this work in that setting?
I agree, we should definitely make them as functions tf compatible (similar to what we have). 
I don't see why chaining, merge ad split transform functions could be a problem for tf computational graph. 
However, we can simply make transform functions in a similar format as--- warp_features_labels in transform.py. We can choose not to combine them in channels.

>     * what about looking at the scikit-learn transform API, would that be useful?
Actually composetranform() is very similar to Pipelines and composite estimators in scikit-learn. 
 Hmm, anything specific that needs to be seen in scikit-learn transform?

>     * what about multimodal inputs and outputs?
Interesting point. But maybe we should elaborate on the question side here and discuss what type of multimodal data are we expecting to process. It would be usefull to design transforms also accordingly, eg. within imaging space, MRIs and CTs have different preprocessing steps.

Aakanksha-Rana commented:
> > we may want to augment as part of a tensorflow graph as opposed to outside
> 
> I agree, we should do this as part of the tensorflow graph, preferably inside a `tf.data.Dataset` operation ([`dataset.map`](https://www.tensorflow.org/api_docs/python/tf/data/Dataset#map)). Those operations can be parallelized, which can be very useful for augmentations that take significant time. I have implemented a few operations with tensorflow to be used in a tf.data dataset (train_dataset.map)
> 
Yes, totally, the transform functionality would be compatible for tf computational graph. They would be present in the dataloader function which is map method of (tf.data.Dataset). Plan is to do very similar to already implemented functions in transform.py and volume.py.

> > what about multimodal inputs and outputs
> 
> multimodal inputs are a good argument for not concatenating features and labels in the channel dimension. I would prefer if we could keep features and labels separate but create transformation compositions that operate on both of them. i wonder what the scikit-learn api has done about this
>
Mentioned in comments from Satra. This is a great point and we should also discuss this:
--what all modalities we might be process, 
--Merging wont be suitable for Multimodal data, but different modalities needs different transformations as well (e.g normalization of MRIs and CTs would be different.). 
--Also, depending on modalities, what all added transformation do we need.  

> edit: regarding the scikit-learn data transform api, the sklearn pipeline expects all transform objects to implement `.fit(x, y=None)`. Perhaps we can use something like this. Then, features and labels are passed to transform objects together, and we can implement `.fit` methods that deal with those appropriately. For example, random rigid transformations would apply to both `x` and `y`, but standardizing data would only apply to `x` and `y` would be passed through unchanged.

Merging and splitting was just an added feature may/may-not be used. We can also go the way it is already in examples like warp_features_labels in transform.py .  We can go like Transform3() in the design for all functions! 

btw, I couldn't find any chaining function ( which also exits in scikit-learn api) in tf framework. Is it so ? Maybe I am missing. something.

satra commented:
perhaps consider: `.transform(x, y=None)` with the init signature something like:

`__ini__(..., transform_both=False, x_params=None, y_params=None)`

as an example if the transform was resample:

`Resample(..., interp=Linear, transform_xy=True, x_params=dict(interp='BSpline'), y_params=dict(interp='NearestNeighbor'))`

where x and y params override any of the default parameters.

Hoda1394 commented:
I will review this.

kaczmarj commented:
thanks @Hoda1394 - i was also thinking about using non-tensorflow implementations for image augmentation, like scikit-image. the downside of going outside of tensorflow is that we will not be able to serialize the operations into a graph, but i the benefits outweigh this limitation. the main benefits are flexibility and ease. we can use scikit-image's implementation of affine transformations, noise, flips, etc.

this might not affect this pull request, but i wanted to write down these thoughts somewhere useful...

satra commented:
@Aakanksha-Rana - is this still a wip?

satra commented:
@Aakanksha-Rana - i've updated the nobrainer repo to use precommit to check for a bunch of things. you will have to fetch your branch and then run pre-commit locally and then push back.
